tools/find_deprecated.py

```diff
@@ -0,0 +1,175 @@
+        self.filename = filename
+        self.deprecations = []
+
+    @staticmethod
+    def is_deprecation_decorator(node):
+        """Check if a node is a deprecation decorator"""
+        return (
+            isinstance(node, ast.Call)
+            and isinstance(node.func, ast.Name)
+            and node.func.id.startswith("deprecate_")
+        )
+
+    def visit_FunctionDef(self, node):  # pylint: disable=invalid-name
+        """Visitor for function declarations"""
+        self.deprecations += [
+            Deprecation(self.filename, d_node, node)
+            for d_node in node.decorator_list
+            if DecoratorVisitor.is_deprecation_decorator(d_node)
+        ]
+        ast.NodeVisitor.generic_visit(self, node)
+
+
+class DeprecationCollection:
+    """
+    A collection of :class:~.Deprecation
+
+    Args:
+        dirname: Directory name that would be checked recursively for deprecations.
+    """
+
+    def __init__(self, dirname):
+        self.dirname = dirname
+        self._deprecations = None
+        self.grouped = OrderedDict()
+
+    @property
+    def deprecations(self):
+        """List of deprecation :class:~.Deprecation"""
+        if not self._deprecations:
+            self.collect_deprecations()
+        return self._deprecations
+
+    def collect_deprecations(self):
+        """Run the :class:~.DecoratorVisitor on `self.dirname`"""
+        self._deprecations = []
+        files = [self.dirname] if self.dirname.is_file() else self.dirname.rglob("*.py")
+        for filename in files:
+            self._deprecations.extend(DeprecationCollection.find_deprecations(filename))
+
+    def group_by(self, attribute_idx):
+        """Group :class:~`.Deprecation` in self.deprecations based on the attribute attribute_idx"""
+        grouped = defaultdict(list)
+        for obj in self.deprecations:
+            grouped[getattr(obj, attribute_idx)].append(obj)
+        for key in sorted(grouped.keys()):
+            self.grouped[key] = grouped[key]
+
+    @staticmethod
+    def find_deprecations(file_name):
+        """Runs the deprecation finder on file_name"""
+        with open(file_name, encoding="utf-8") as fp:
+            code = fp.read()
+        mod = ast.parse(code, file_name)
+        decorator_visitor = DecoratorVisitor(file_name)
+        decorator_visitor.visit(mod)
+        return decorator_visitor.deprecations
+
+
+if __name__ == "__main__":
+    collection = DeprecationCollection(Path(__file__).joinpath("..", "..", "qiskit").resolve())
+    # collection.collect_deprecations()
+    collection.group_by("since")
+
+    DATA_JSON = LAST_TIME_MINOR = DETAILS = None
+    try:
+        DATA_JSON = requests.get("https://pypi.org/pypi/qiskit-terra/json", timeout=5).json()
+        LAST_MINOR = ".".join(DATA_JSON["info"]["version"].split(".")[:2])
+        LAST_TIME_MINOR = datetime.fromisoformat(
+            DATA_JSON["releases"][f"{LAST_MINOR}.0"][0]["upload_time"]
+        )
```

Comment:
These don't need to be in the in the `try` block. It's useful to have the `try` block be as small as possible to make it clear what is exception-raising and what is not.

Reply:
done in 68f19da0f8ef807898d05773654d47f47477df3e